command/agent/auth/auth.go

```diff
@@ -311,3 +326,14 @@ func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) error {
 		}
 	}
 }
+
+// calculateBackoff determines a new backoff duration that is roughly twice
+// the previous value, capped to a max value, with a measure of randomness.
+func calculateBackoff(previous, max time.Duration) time.Duration {
+	maxBackoff := math.Min(2*previous.Seconds(), max.Seconds())
+	backoff := 0.75*maxBackoff + 0.25*rand.Float64()*maxBackoff
+
+	// Scale to ms to avoid truncation errors for < 2s values, which can
+	// leave the backoff stuck at 1s on successive calls.
+	return time.Duration(backoff*1000) * time.Millisecond
```

Comment:
heh... I mainly wanted to use `math.Min` and avoid `float64` conversions. Not sure about nanoseconds, since there might(?) be precision issues when converting to float, and integral math starts getting ugly (e.g. what to do about that rand call). Here is a millisecond version. Any clearer?

```
func calculateBackoff(previous, max time.Duration) time.Duration {
	maxBackoff := 2 * previous.Milliseconds()
	if maxBackoff > max.Milliseconds() {
		maxBackoff = max.Milliseconds()
	}

	backoff := 0.75*float64(maxBackoff) + 0.25*rand.Float64()*float64(maxBackoff)

	return time.Duration(backoff) * time.Millisecond
}
```

Reply:
Could you operate in nanoseconds internally, i.e. `maxBackoff := math.Min(2*previous.Nanoseconds(), max.Nanoseconds())` and return the duration in ns?

Reply:
@calvn I just updated my previous example with something cleaner.

Reply:
Nice, I think that reads really well, although I don't think the `int64()` cast for 4 is required.

Reply:
Oops, yep. Removed.

Reply:
Updated to this in 7f70409.